File: mailtidy/data/reports.py
"""报告生成：把 ``AgentPlan`` / ``ExecutionResult`` 转成人类可读输出。

为什么独立成一个模块：
- 让"决策逻辑"和"展示逻辑"严格分离，未来要换 HTML / Slack Block Kit /
  Telegram Markdown 等格式时，只动这里。
- 测试更简单：只校验输入数据 -> 字符串映射，不需要 mock IO。

当前所有输出都是 Markdown，对终端 / Slack / Notion 等场景兼容性最好。
"""

from __future__ import annotations

import csv
from io import StringIO

from mailtidy.data.models import AgentPlan, Category, EmailMessage, ExecutionResult


def cleanup_report(plan: AgentPlan, result: ExecutionResult, messages: list[EmailMessage], newsletter_summary: str) -> str:
    """生成"清理完成"报告。"""
    by_id = {message.id: message for message in messages}
    attention = [
        judgment
        for judgment in plan.judgments
        if judgment.category in {Category.IMPORTANT, Category.ACTIONABLE}
    ]
    lines = [
        "# MailTidy Cleanup Report",
        "",
        f"- Processed: {result.processed}",
        f"- Archived: {result.archived}",
        f"- Labeled: {result.labeled}",
        f"- Starred: {result.starred}",
        f"- Marked read: {result.marked_read}",
        f"- Estimated cost: development heuristic mode ($0.00)",
        "",
        "## Needs Your Attention",
    ]
    if attention:
        for judgment in attention:
            message = by_id[judgment.email_id]
            lines.append(f"- {message.sender}: {message.subject} - {judgment.reason}")
    else:
        lines.append("- None")
    lines.extend(["", "## Newsletter Summary", newsletter_summary])
    if plan.human_prompts:
        lines.extend(["", "## Confirmation Needed", *[f"- {prompt}" for prompt in plan.human_prompts]])
    return "\n".join(lines)


def daily_brief(plan: AgentPlan, messages: list[EmailMessage]) -> str:
    """生成"早晨 2 分钟可读"的日报。"""
    by_id = {message.id: message for message in messages}
    groups = {
        "Urgent Today": [j for j in plan.judgments if j.urgency >= 4],
        "Important This Week": [j for j in plan.judgments if 2 <= j.urgency < 4],
        "FYI": [j for j in plan.judgments if j.urgency < 2],
    }
    lines = ["# MailTidy Daily Brief", ""]
    for title, judgments in groups.items():
        lines.append(f"## {title}")
        if not judgments:
            lines.append("- None")
        for judgment in judgments:
            message = by_id[judgment.email_id]
            lines.append(f"- [{message.sender}] {message.subject} - {message.snippet} - {judgment.action_suggestion}")
        lines.append("")
    lines.append(f"Unread: {len(messages)}; likely needs reply: {sum(1 for j in plan.judgments if j.category == Category.ACTIONABLE)}")
    return "\n".join(lines)


def subscriptions_markdown(rows: list[dict[str, object]]) -> str:
    """订阅扫描的 Markdown 输出。"""
    total = sum(float(row.get("monthly_amount", 0)) for row in rows)
    lines = [
        "# Subscription Scan",
        "",
        f"You have {len(rows)} likely active subscription(s), about ${total:.2f}/month.",
        "",
        "| Service | Monthly | Category | Last Charge | Plan |",
        "| --- | ---: | --- | --- | --- |",
    ]
    for row in rows:
        lines.append(
            f"| {row['service_name']} | ${float(row['monthly_amount']):.2f} | "
            f"{row['category']} | {row['last_charge_date']} | {row['plan_name']} |"
        )
    return "\n".join(lines)


def subscriptions_csv(rows: list[dict[str, object]]) -> str:
    """订阅扫描的 CSV 输出，方便用户导入 Excel / Google Sheets 二次分析。"""
    output = StringIO()
    fieldnames = ["service_name", "monthly_amount", "currency", "billing_cycle", "last_charge_date", "plan_name", "category"]
    writer = csv.DictWriter(output, fieldnames=fieldnames)
    writer.writeheader()
    writer.writerows({key: row.get(key, "") for key in fieldnames} for row in rows)
    return output.getvalue()


__all__ = [
    "cleanup_report",
    "daily_brief",
    "subscriptions_csv",
    "subscriptions_markdown",
]

File: src/data/reports.ts
/**
 * 报告生成：把 AgentPlan / ExecutionResult 转成人类可读输出。
 *
 * 让"决策逻辑"和"展示逻辑"严格分离，未来要换 HTML / Slack Block Kit /
 * Telegram Markdown 等格式时只动这里。
 */

import { Category, type AgentPlan, type EmailMessage, type ExecutionResult } from "./models.js";

export interface SubscriptionRow {
  serviceName: string;
  monthlyAmount: number;
  currency: string;
  billingCycle: string;
  lastChargeDate: string;
  planName: string;
  category: string;
  unsubscribeLink?: string;
}

export function cleanupReport(
  plan: AgentPlan,
  result: ExecutionResult,
  messages: EmailMessage[],
  newsletterSummary: string,
): string {
  const byId = new Map(messages.map((m) => [m.id, m]));
  const attention = plan.judgments.filter(
    (j) => j.category === Category.IMPORTANT || j.category === Category.ACTIONABLE,
  );
  const lines = [
    "# MailTidy Cleanup Report",
    "",
    `- Processed: ${result.processed}`,
    `- Archived: ${result.archived}`,
    `- Labeled: ${result.labeled}`,
    `- Starred: ${result.starred}`,
    `- Marked read: ${result.markedRead}`,
    `- Estimated cost: development heuristic mode ($0.00)`,
    "",
    "## Needs Your Attention",
  ];
  if (attention.length === 0) {
    lines.push("- None");
  } else {
    for (const j of attention) {
      const m = byId.get(j.emailId);
      if (!m) continue;
      lines.push(`- ${m.sender}: ${m.subject} - ${j.reason}`);
    }
  }
  lines.push("", "## Newsletter Summary", newsletterSummary);
  if (plan.humanPrompts.length > 0) {
    lines.push("", "## Confirmation Needed", ...plan.humanPrompts.map((p) => `- ${p}`));
  }
  return lines.join("\n");
}

export function dailyBrief(plan: AgentPlan, messages: EmailMessage[]): string {
  const byId = new Map(messages.map((m) => [m.id, m]));
  const groups: Array<[string, typeof plan.judgments]> = [
    ["Urgent Today", plan.judgments.filter((j) => j.urgency >= 4)],
    ["Important This Week", plan.judgments.filter((j) => j.urgency >= 2 && j.urgency < 4)],
    ["FYI", plan.judgments.filter((j) => j.urgency < 2)],
  ];
  const lines = ["# MailTidy Daily Brief", ""];
  for (const [title, judgments] of groups) {
    lines.push(`## ${title}`);
    if (judgments.length === 0) lines.push("- None");
    for (const j of judgments) {
      const m = byId.get(j.emailId);
      if (!m) continue;
      lines.push(`- [${m.sender}] ${m.subject} - ${m.snippet} - ${j.actionSuggestion}`);
    }
    lines.push("");
  }
  const replies = plan.judgments.filter((j) => j.category === Category.ACTIONABLE).length;
  lines.push(`Unread: ${messages.length}; likely needs reply: ${replies}`);
  return lines.join("\n");
}

export function subscriptionsMarkdown(rows: SubscriptionRow[]): string {
  const total = rows.reduce((sum, r) => sum + r.monthlyAmount, 0);
  const lines = [
    "# Subscription Scan",
    "",
    `You have ${rows.length} likely active subscription(s), about $${total.toFixed(2)}/month.`,
    "",
    "| Service | Monthly | Category | Last Charge | Plan |",
    "| --- | ---: | --- | --- | --- |",
  ];
  for (const r of rows) {
    lines.push(
      `| ${r.serviceName} | $${r.monthlyAmount.toFixed(2)} | ${r.category} | ${r.lastChargeDate} | ${r.planName} |`,
    );
  }
  return lines.join("\n");
}

export function subscriptionsCsv(rows: SubscriptionRow[]): string {
  const fieldnames = [
    "service_name",
    "monthly_amount",
    "currency",
    "billing_cycle",
    "last_charge_date",
    "plan_name",
    "category",
  ];
  const out: string[] = [fieldnames.join(",")];
  for (const r of rows) {
    out.push(
      [
        r.serviceName,
        r.monthlyAmount.toFixed(2),
        r.currency,
        r.billingCycle,
        r.lastChargeDate,
        r.planName,
        r.category,
      ]
        .map((v) => (typeof v === "string" && v.includes(",") ? `"${v}"` : String(v)))
        .join(","),
    );
  }
  return out.join("\n") + "\n";
}
